src/Data/Array/Polarized.hs

```diff
@@ -1,6 +1,103 @@
 {-# LANGUAGE LinearTypes #-}
 {-# LANGUAGE NoImplicitPrelude #-}
 
+-- | This module provides converters and documentation for polarized arrays.
+--
+-- There are two polarities or \"directions\": push arrays and pull arrays.
+--
+-- You should use push an pull arrays if you want to explicity control
+-- when GHC allocates space in the garbage-collector heap for arrays.
+-- The point of doing this is to reduce the amount of memory used
+-- and improve the performance of your program.
+--
+-- The performance of your program is improved because linear types enable the
+-- API which does not allow operations on push or pull arrays that cannot be
+-- fused. That is, for some subtle and complicated reasons, we can be sure that
+-- only one function allocates memory (and it's called 'alloc') and all other
+-- compositions do not allocate memory.
+--
+-- == How to use push and pull arrays
+--
+-- __The trifecta of controlling allocation:__
+--
+-- \[ \texttt{Vector a} \]
+-- \[ \text{fromVector}~~ { \Huge \swarrow  ~~~~ \nwarrow } ~~~ \text{alloc} \]
+-- \[ \texttt{Pull a} {\Huge ~~~ \longrightarrow ~~~} \texttt{Push a}\]
+-- \[ \text{transfer}\]
+--
+-- Pull arrays are good as arguments to functions, because they are easy to
+-- \"pull\" from, map over and zip with.
+--
+-- Push arrays are good as values returned from functions, and are easy to
+-- \"write\" to and modify.
+--
+-- A typical application starts with concrete lists as @Vector@s (which are
+-- inputs), goes to @Pull@ arrays, then to @Push@ arrays and ends with
+-- @Vector@s (which are outputs).  All the computations inbetween the
+-- "fromVector" and "alloc" are fused and no memory from the GC heap is
+-- allocated when they are executed.
+--
+-- Such a combination of functions that controls fusion is called 
+-- a __fusion pipeline__.
+--
+-- == A toy example
+--
+-- > import Data.Array.Polarized
+-- > import qualified Data.Array.Polarized.Push as Push
+-- > import qualified Data.Array.Polarized.Pull as Pull
+-- > import qualified Data.Functor.Linear as Linear
+-- > import Data.Vector (Vector, (!), fromList)
+-- > import Prelude
+-- >
+-- > type Pull a = Pull.Array a
+-- > type Push a = Push.Array a
+-- >
+-- > pipeline :: Vector Int -> Vector Int -> Vector Int
+-- > pipeline coeff =
```

Comment:
Did you try to typecheck this example? It doesn't look correct to me.

Reply:
@Divesh-Otwani I can recommend doctests for this sort of thing. The setup is fiddly, but when it works, it's nice.

Example introduction of doctests into a codebase: https://github.com/idontgetoutmuch/random/pull/34/files

The doctest syntax is described here: https://github.com/sol/doctest

Reply:
Thanks -- I did type check it but random stuff does go wrong so a doctest is really useful.